sdk/Lusid.Sdk.Tests/Utilities/PollyApiRetryHandlerTests.cs

```diff
@@ -0,0 +1,409 @@
+
+            var testApiConfig = TestLusidApiFactoryBuilder.CreateApiConfiguration(null);
+            testApiConfig.ApiUrl = ListenerUriPrefix;
+
+            _apiFactory = new LusidApiFactory(testApiConfig);
+
+            _httpListener.Start();
+        }
+
+        [Test]
+        public void CallGetPortfoliosApi_WhenHttpStatusIs400AndRetryConditionIsNotSatisfied_ThrowsApiExceptionWithoutRetry()
+        {
+            // It should not retry on codes >400 by default, unless specified
+            const int expectedStatusCode = 400;
+            // Add the next response returned by api
+            AddMockHttpResponseToQueue(_httpListener, statusCode: expectedStatusCode, responseContent: "{\"some\": \"JsonResponseHere\"}");
+            var retryCount = 0;
+            RetryConfiguration.RetryPolicy =
+                Policy.Wrap(
+                    PollyApiRetryHandler.DefaultFallbackPolicy,
+                    Policy
+                        .HandleResult<IRestResponse>(PollyApiRetryHandler.GetInternalExceptionRetryCondition)
+                        .Retry(retryCount: 3, onRetry: (result, i) => retryCount++)
+                );
+        
+            var exception = Assert.Throws<ApiException>(
+                () => _apiFactory.Api<IPortfoliosApi>().GetPortfolio("any", "any")
+            );
+
+            Assert.That(exception.ErrorContent, Is.EqualTo("{\"some\": \"JsonResponseHere\"}"));
+            Assert.That(exception.ErrorCode, Is.EqualTo(expectedStatusCode));
+            Assert.That(retryCount, Is.EqualTo(0));
+        }
+
+        [Test]
+        public void CallGetPortfoliosApi_WhenHttpStatusIs200AndRetryConditionIsNotSatisfied_NoPollyRetryIsTriggered()
+        {
+            // It should do nothing when response code is 200
+            const int expectedStatusCode = 200;
+            // Add the next response returned by api
+            AddMockHttpResponseToQueue(_httpListener, expectedStatusCode, responseContent: "{\"some\": \"JsonResponseHere\"}");
+            var retryCount = 0;
+            RetryConfiguration.RetryPolicy =
+                Policy.Wrap(
+                    PollyApiRetryHandler.DefaultFallbackPolicy,
+                    Policy
+                        .HandleResult<IRestResponse>(PollyApiRetryHandler.GetInternalExceptionRetryCondition)
+                        .Retry(retryCount: 3,
+                            onRetry: (result, i) => retryCount++)
+                );
+
+            // TODO: This will still fail at deserialization, which currently is not handled and returns null.
+            // Will be done under a different PR. Response string will also be changed.
+            var sdkResponse = _apiFactory.Api<IPortfoliosApi>().GetPortfolio("any", "any");
+
+            Assert.That(retryCount, Is.EqualTo(0));
+        }
+
+        [Test]
+        public void CallGetPortfoliosApi_WhenApiResponseStatusCodeSatisfiesRetryCriteria_ExceedsPollyRetriesAndThrows()
```

Comment:
What is this test checking? Is it asserting that the `PollyApiRetryHandler.DefaultFallbackPolicy` works as expected? Is it asserting that you can write a Polly retry policy that retries on a certain status code? If so what does that demonstrate? You seem to be testing the functionality of Polly here rather than your own RetryHandler that you've created.

Reply:
This is something absolutely outrageous that I've encountered. If there is no fallback policy, then after retries have exceeded, then `policyResult.Outcome` becomes  `OutcomeType.Failure`, here: 
https://github.com/finbourne/lusid-sdk-csharp-preview/blob/master/sdk/Lusid.Sdk/Client/ApiClient.cs#L494
which as a result constructs a response with absolutely NO INFORMATION to be used and to display to the user.

What I'm testing is that whenever we perhaps change any any default templates, upgrade the SDK or anything, the behaviour is as expected, that is, if retries have been exceeded, we then throw the error that would also be thrown if there were no retries.

Reply:
Ok